### ISSUE TITLE
Should GPUDevice.destroy detach all buffer mapping ArrayBuffers?

### ISSUE THREAD

**Issue by Kangz** [gpuweb/gpuweb#2545]
Right now in the specification calling `GPUDevice.destroy` mostly just makes the device invalid. However that's an explicit signal from the application that it wants to stop using that WebGPU device, so should it also implicitly call `GPUBuffer.destroy` for all `GPUBuffers`? This way `ArrayBuffer` from that device can be reclaimed early as well as their backing stores.

**Comment by lokokung** [comment]
Destroy currently calls the internal buffer.destroy already. I'm not entirely sure if that achieves what the desired result here is, but this is true for all device created objects as well, i.e. textures.

**Comment by Kangz** [comment]
Where in the specification is that written? I can't find it.

**Comment by kainino0x** [comment]
I think all we have is a non-normative note saying that the implementation can freely destroy everything:
https://gpuweb.github.io/gpuweb/#dom-gpudevice-destroy

I don't think it's really necessary to actually specify that .destroy is called on all the textures/buffers/query sets. That operation exists basically to detach the resource so validation prevents you from using it again, but the device being lost also [prevents you from using them again](https://gpuweb.github.io/gpuweb/#abstract-opdef-valid-to-use-with). Also, it only applies to 3 types of objects since there are no destroy methods on everything else.

**Comment by Kangz** [comment]
The question is about the `DOMArrayBuffer` returned from `getMappedRange`. When `GPUBuffer.destroy` is called they are detached. Is that the same for `GPUDevice.destroy` or not? Since it is observable, it must be specified one way or another. (I think it should be specified that they get detached)

**Comment by lokokung** [comment]
In dawn, when device.destroy is called, the buffers.destroy is also called. However, looking at the Chromium bindings, it looks like it does an extra operation on buffer.destroy: https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/modules/webgpu/gpu_buffer.cc;l=139;bpv=1;bpt=0. 

I suspect that's what you mean, so in that regard I think the answer is no at the moment. If we want this feature though, it might be tricky since this needs to be done at the binding level which means maybe we need to keep track of created objects in blink also? (Or maybe I am misunderstanding something?)

**Comment by kainino0x** [comment]
> The question is about the `DOMArrayBuffer` returned from `getMappedRange`

Oh, sorry, I totally didn't read closely enough. It would make sense to detach all the mapped buffers on destroy() - however I wouldn't want to do it on other kinds of device loss because those are delivered asynchronously.

As a result of that it should probably be allowed to device.destroy() in the callback after receiving `device.lost`, and have that clean up any ArrayBuffers. (No concerns about that recursing, because `device.lost` can only ever resolve once.)

> If we want this feature though, it might be tricky since this needs to be done at the binding level which means maybe we need to keep track of created objects in blink also? (Or maybe I am misunderstanding something?)

Yeah, we would need to keep track at the "content timeline" (Blink) layer, but at least it's only for GPUBuffers and not everything else. @austinEng is currently working on exactly that in Blink. I think the spec should do essentially the same thing, track on the GPUDevice all GPUBuffers that might have ArrayBuffers attached to them so we can clean them up.

**Comment by kainino0x** [comment]
Editor's agree on doing this (as I proposed).

> I wouldn't want to do it on other kinds of device loss because those are delivered asynchronously

Relates to #1629

**Comment by greggman** [comment]
Sorry if I'm over thinking this. Does this matter for multi-threaded stuff. If I have a device shared with a worker and I call `device.destory()` in the main thread the proposal is that DOMArrayBuffers in the worker get immediately affected?

**Comment by Kangz** [comment]
While we don't have support for multiple workers yet, we don't want to design ourselves in a corner. We can't detach ArrayBuffers between worker, so we probably need to think about this more.

**Comment by kainino0x** [comment]
> Sorry if I'm over thinking this. Does this matter for multi-threaded stuff. If I have a device shared with a worker and I call `device.destory()` in the main thread the proposal is that DOMArrayBuffers in the worker get immediately affected?

Good point. We could probably have destroy() only affect arraybuffers on the same thread, and have other threads see it as a device loss? But probably better to just not detach them on destroy at all, and force buffers to be unmapped.

GPUBuffer.destroy from a thread has the same problem. I think unmap will only be valid from the thread that holds the map. So we should probably remove the unmap() call from inside GPUBuffer.destroy().

**Comment by kainino0x** [comment]
OTOH maybe doing the unmap-if-same-thread in both cases is nicer for the "common" case of single-threaded applications?

**Comment by kainino0x** [comment]
Editors: two alternatives:

1. Never automatically unmap, you have to do it manually (or just let the ArrayBuffer objects get garbage collected). Con: Mappings likely to leak until they get GC'd. Maybe not that bad?
2. Don't unmap on device destroy, instead unmap (_on all threads_) in the task where the device.lost promise gets resolved (maybe only if the device was lost _due_ to destroy). Do something equivalent for buffer destroy. Con: Still can be disruptive to your other threads. Diverging the two types of device loss maybe not ideal.

**Comment by kainino0x** [comment]
Editors meeting: The more I mull on this, the more option 1 is the only one that makes sense: it further separates buffer mapping from the rest of the device so that it doesn't have weird interactions.

- In the spirit of #1629, spontaneous device loss shouldn't detach ArrayBuffers because it could "crash" the application (throw a JS exception). With threads, destroying the device on one thread is very much like spontaneous device loss on another thread.
- Ongoing async work shouldn't have to constantly check whether the device is lost (or even just destroyed) in order to keep operating.
- It seems annoying if ongoing tasks (like async work) behave differently on device destruction depending on which thread the destruction came from.
- Continued access to mapped buffers after device destroy (use-after-free) isn't any more of a concern than after a real device loss.

It's unfortunate that device.destroy wouldn't be a comprehensive way to clean up, but since mapped buffers are so tied to JS execution (unlike other WebGPU operations which are async), I don't think it works out nicely. If applications want to clean up more promptly, they can implement their own logic for unmapping buffers from destroyed devices.

@Kangz what do you think of this?

---

This also raises the related point of what happens to mappings on _buffer_ destroy. I would propose doing the same thing: just don't unmap on destroy (option 1). This is somewhat less necessary since buffer destruction is less global, but it still interacts with threads in the same way. An equivalent of option 2 would be to unmap only if the destroy is on the same thread, otherwise leave mapped, but this is too exceptional and invites use-after-free bugs. An option 3 would be to disallow destroy while a buffer is mapped (on any thread), but this seems like an unnecessary special case.

**Comment by kainino0x** [comment]
@Kangz and I discussed the original proposal (I'll call it option 4) and ironed it out as follows:

- device.destroy() immediately unmaps all buffers **on the same thread**
- buffers on other threads remain mapped
- you can device.destroy() on each thread separately to clean up the maps on that thread (destroying multiple times is OK)
- buffer.destroy() can probably work the same way

(We briefly discussed another option where you could call something like device.close() to remove a device from a thread and unmap everything (and destroy may or may not call this) but I think that doesn't work well because it would introduce a new state checked by EVERY operation on every object from that device on that thread.)
